File: ml6-d-svn/trunk/transformations/MC/matrix.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#define R2D 180.0/M_PI
#define D2R M_PI/180.0
#include "matrix.h"


/*-------------- struct matrix *new_matrix() --------------
Inputs:  int rows
         int cols 
Returns: 

Once allocated, access the matrix as follows:
m->m[r][c]=something;
if (m->lastcol)... 
*/

struct matrix *new_matrix(int rows, int cols) {
  double **tmp;
  int i;
  struct matrix *m;

  tmp = (double **)malloc(rows * sizeof(double *));
  for (i=0;i<rows;i++) {
      tmp[i]=(double *)malloc(cols * sizeof(double));
    }

  m=(struct matrix *)malloc(sizeof(struct matrix));
  m->m=tmp;
  m->rows = rows;
  m->cols = cols;
  m->lastcol = cols;

  return m;
}


/*-------------- void free_matrix() --------------
Inputs:  struct matrix *m 
Returns: 

1. free individual rows
2. free array holding row pointers
3. free actual matrix
*/
void free_matrix(struct matrix *m) {

  int i;
  for (i=0;i<m->rows;i++) {
      free(m->m[i]);
    }
  free(m->m);
  free(m);
}


/*======== void grow_matrix() ==========
Inputs:  struct matrix *m
         int newcols 
Returns: 

Reallocates the memory for m->m such that it now has
newcols number of columns
====================*/
void grow_matrix(struct matrix *m, int newcols) {
  int i;
  for (i=0;i<m->rows;i++) {
      m->m[i] = realloc(m->m[i],newcols*sizeof(double));
    }
  m->cols = newcols;
}


/*-------------- void print_matrix() --------------
Inputs:  struct matrix *m 
Returns: 

print the matrix
*/
void print_matrix(struct matrix *m) {
int xx=m->rows;
int yy=m->cols;
int x;
int y;


for(x=0;x<xx;++x){
	printf("\n");
	for(y=0;y<yy;++y){
		printf("|%f|",m->m[x][y]); 
	}
}

}

/*-------------- void ident() --------------
Inputs:  struct matrix *m <-- assumes m is a square matrix
Returns: 

turns m in to an identity matrix
*/
void ident(struct matrix *m) {
fprintf(stderr,"calling IDENT\n");
int xx=m->rows;
int yy=m->cols;
int x;
int y;


for(x=0;x<xx;++x){
	//printf("\n");
	for(y=0;y<yy;++y){
		m->m[x][y]=0;		
	}
}

for(x=0;x<xx;++x){m->m[x][x]=1;}
}


/*-------------- void scalar_mult() --------------
Inputs:  double x
         struct matrix *m 
Returns: 

multiply each element of m by x
*/
void scalar_mult(double xd, struct matrix *m) {
//fprintf(stderr,"calling SCALAR\n");
int xx=m->rows;
int yy=m->cols;
int x;
int y;


for(x=0;x<xx;++x){
	for(y=0;y<yy;++y){
		m->m[x][y]=m->m[x][y]*xd;
			
	}
}

}


/*-------------- void matrix_mult() --------------
Inputs:  struct matrix *a
         struct matrix *b 
Returns: 

a*b -> b
*/

void matrix_mult(struct matrix *a, struct matrix *b) {
//fprintf(stderr,"calling DOTMULT\n");

struct matrix * f;
f=new_matrix(a->rows,b->cols);
//fprintf(stderr,"A:(%d,%d) \nB:(%d,%d)\nF:(%d,%d)",a->rows,a->cols,b->rows,b->cols,f->rows,f->cols);
int x;
int y;
int z;
//fprintf(stderr,"MULTIPLYING");
for(y=0;y<(f->cols);++y){
	for(x=0;x<(f->rows);++x){
		f->m[x][y]=0;
		for(z=0;z<(a->cols);++z){	
			//fprintf(stderr,"X:%d Y:%d Z:%d \n",x,y,z);	
			f->m[x][y]=f->m[x][y]+((a->m[x][z])*(b->m[z][y]));
		}
	}
}
//fprintf(stderr,"FREEING");
copy_matrix(f,b);
free_matrix(f);
}


/*-------------- void copy_matrix() --------------
Inputs:  struct matrix *a
         struct matrix *b 
Returns: 

copy matrix a to matrix b
*/
void copy_matrix(struct matrix *a, struct matrix *b) {
//fprintf(stderr,"calling MATCOP\n");
int xx=a->rows;
int yy=a->cols;
int x;
int y;


for(x=0;x<xx;++x){
	for(y=0;y<yy;++y){
		b->m[x][y]=a->m[x][y];
	}
}

}

/*======== struct matrix * make_translate() ==========
Inputs:  double x
         double y
         double z 
Returns: The translation matrix created using x, y and z 
as the translation offsets.
====================*/
struct matrix * make_translate(double x, double y, double z) {
  struct matrix *a;
  a = new_matrix(4,4);
  ident(a);
  a->m[0][3] = x;
  a->m[1][3] = y;
  a->m[2][3] = z;
  return a;
}

/*======== struct matrix * make_scale() ==========
Inputs:  double x
         double y
         double z 
Returns: The translation matrix creates using x, y and z
as the scale factors
====================*/
struct matrix * make_scale(double x, double y, double z) {
  struct matrix *a;
  a = new_matrix(4,4);
  ident(a);
  a->m[0][0] = x;
  a->m[1][1] = y;
  a->m[2][2] = z;
  return a;
}

/*======== struct matrix * make_rotX() ==========
Inputs:  double theta

Returns: The rotation matrix created using theta as the 
angle of rotation and X as the axis of rotation.
====================*/
struct matrix * make_rotX(double theta) {
  struct matrix *a;
  a = new_matrix(4,4);
  ident(a);
  a->m[1][1] = cos(theta * D2R);
  a->m[2][1] = -1 * sin(theta * D2R);
  a->m[1][2] = sin(theta * D2R);
  a->m[2][2] = cos(theta * D2R);
  return a;
}

/*======== struct matrix * make_rotY() ==========
Inputs:  double theta
         char c 
Returns: The rotation matrix created using theta as the 
angle of rotation and Y as the axis of rotation.
====================*/
struct matrix * make_rotY(double theta) {
  struct matrix *a;
  a = new_matrix(4,4);
  ident(a);
  a->m[0][0] = cos(theta * D2R);
  a->m[2][0] = sin(theta * D2R);
  a->m[0][2] = -1 * sin(theta * D2R);
  a->m[2][2] = cos(theta * D2R);
  return a;
}

/*======== struct matrix * make_rotZ() ==========
Inputs:  double theta
         char c 
Returns: The rotation matrix created using theta as the 
angle of rotation and Z as the axis of rotation.
====================*/
struct matrix * make_rotZ(double theta) {
  struct matrix *a;
  a = new_matrix(4,4);
  ident(a);
  a->m[0][0] = cos(theta * D2R);
  a->m[1][0] = -1 * sin(theta * D2R);
  a->m[0][1] = sin(theta * D2R);
  a->m[1][1] = cos(theta * D2R);
  return a;
}

File: ml6-d-svn/trunk/transformations/Michael_Hammer/Matrix.java
/*========== Matrix.java ==========
  Matrix will hold a 2-d array of doubles and have a default size of 4x4.
  Handles basic matrix maintenence and math.
  Creates transformation matricies for tralation, scale and rotate
=========================*/

/****
  * Sample Matrix (4 rows X 5 columns)
  * 3 5 6 7 5
  * 4 2 1 3 2
  * 0 9 8 3 2
  * 1 3 5 6 8
  *
  * m[0][0] m[0][1] m[0][2] m[0][3] m[0][4]
  * m[1][0] m[1][1] m[1][2] m[1][3] m[1][4]
  * m[2][0] m[2][1] m[2][2] m[2][3] m[2][4]
  * m[3][0] m[3][1] m[3][2] m[3][3] m[3][4]
  *
  * */

import java.io.*;
import java.util.*;

public class Matrix
{
  protected double[][] m;
  public static final int DEFAULT_SIZE = 4;
 
  /*===========Constructors================
      Default constructor creates a 4x4 matrix
      Second constructor creates a 4xN matrix
      */
  public Matrix()
  {
    m = new double[DEFAULT_SIZE][DEFAULT_SIZE];
  }
 
  public Matrix(int c)
  {
    m = new double[DEFAULT_SIZE][c];
  }
 
  public double[][] getMatrix()
  {
    return m;
  }
 
  public void setMatrix(double[][] matrixArray)
  {
    m = matrixArray;
  }

  public void setval(int r, int c, double psv)
  {
      if (r < DEFAULT_SIZE && c < m[0].length)
      {
          m[r][c] = psv;
      }
  }



  /*===========grow================
      Increase the number of columns in a matrix by 10
      You can change the growth factor as you see fit
    */
  public void grow()
  {
    double[][] n = new double[m.length][m[0].length + 10];
    for (int r=0; r<m.length; r++)
    {
      for (int c=0; c<m[r].length; c++)
      {
        n[r][c] = m[r][c];
      }
    }
    m = n;
  }

  /*======== public void clear() ==========
   Inputs: 
   Returns:
   Sets every entry in the matrix to 0
   ====================*/
  public void clear()
  {
    for (int i=0; i<m.length; i++)
    {
      for (int j=0; j<m[i].length; j++)
      {
        m[i][j] = 0;
      }
    }
  }

  /*===========ident================
   Turns this matrix into the indentity matrix
   You may assume the calling Matrix is square
   */
  public void makeIdentity()
  {
    this.clear();
    int row = m.length;
    int col = m[row].length;
    for (int i = 0; i < row; i++)
    {
      m[i][i] = 1;
    }
  }

  /*===========scalarMult================
   Inputs:  double x
  
   multiply each element of the calling matrix by x
   */
  public void scalarMult( int s )
  {
    int psv = s;
    for (int i = 0; i < m.length; i++)
    {
      for (int j = 0; j < m[0].length; j++)
      {
        m[i][j] = psv * m[i][j];
      }
    }
  } 

  /*===========matrixMult================
   Multply matrix n by the calling matrix, modify
   the calling matrix to store the result.
  
   eg.
   In the call a.matrixMult(n), n will remain the same
   and a will now be the product of n * a
   */
  public void matrixMult( Matrix n )
  {
    double[][] q = n.getMatrix();
    double[][] psv = new double[m.length][q[0].length];
   
    for (int i = 0; i < m.length; i++) // for each row in M
    {
      for (int j = 0; j < q[0].length; j++) // for each column in N
      {
        for (int a = 0; a < m[0].length; a++) // for each column in M = for each row in N
        {
          psv[i][j] += m[i][a] * q[a][j];
        }
      }
    }
   
    this.setMatrix(psv);
  }
  
  /*===========copy================
   Create and return new matrix that is a duplicate
   of the calling matrix
   */
  public Matrix copy()
  {
    Matrix q = new Matrix(m[0].length);
    double[][] psv = m;
   
   /* for (int i = 0; i < m.length; i++)
    {
      for (int j = 0; j < m[0].length; j++)
      {
        psv[i][j] = m[i][j];
      }
    }*/
   
    q.setMatrix(psv);
    return q;
  }

  /*===========toString================
   Crate a readable String representation of the
   calling matrix.
   */
  public String toString()
  {
    String s = "";
   
    for (int i = 0; i < m.length; i++)
    {
      for (int j = 0; j < m[0].length; j++)
      {
        s = s + m[i][j] + " ";
      }
      s = s + "\n";
    }
   
    return s;
  }

  /*===========MakeTranslate================
      Turns the calling matrix into the appropriate
      translation matrix using x, y, and z as the translation
      offsets.
     */
    public void setTranslate(int x, int y, int z)
    {
    this.makeIdentity();
    this.setval(0, DEFAULT_SIZE, x);
    this.setval(1, DEFAULT_SIZE, y);
    this.setval(2, DEFAULT_SIZE, z);
    }
   
    /*===========MakeScale================
      Turns the calling matrix into the appropriate scale
      matrix using x, y and z as the scale factors.
     */
    public void setScale(int x, int y, int z)
    {
    this.makeIdentity();
    this.setval(0, 0, x);
    this.setval(1, 1, y);
    this.setval(2, 2, z);
    }

    /*=========== MakeRotX ================
      Turns the calling matrix into the appropriate rotation
      matrix using theta as the angle of rotation and X
      as the axis of rotation.
    */
    public void setRotateX(double a)
    {
    double r = Math.PI*a / 90;
    this.makeIdentity();
    this.setval(1, 1, Math.cos(r));
    this.setval(1, 2, -Math.sin(r));
    this.setval(2, 1, Math.sin(r));
    this.setval(2, 2, Math.cos(r));
    }

    /*=========== MakeRotY ================
      Turns the calling matrix into the appropriate rotation
      matrix using theta as the angle of rotation and Y
      as the axis of rotation.
    */
    public void setRotateY(double a)
    {
    double r = Math.PI*a / 90;
    this.makeIdentity();
    this.setval(0, 0, Math.cos(r));
    this.setval(0, 2, -Math.sin(r));
    this.setval(2, 0, Math.sin(r));
    this.setval(2, 2, Math.cos(r));
    }

    /*=========== MakeRotZ ================
      Turns the calling matrix into the appropriate rotation
      matrix using theta as the angle of rotation and axis
      as the axis of rotation.
    */
    public void setRotateZ(double a, char axis)
    {
    double r = Math.PI*a / 90;
    this.makeIdentity();
    this.setval(0, 0, Math.cos(r));
    this.setval(0, 1, -Math.sin(r));
    this.setval(1, 0, Math.sin(r));
    this.setval(1, 1, Math.cos(r));
    }

 
}